src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+
+        // (I) Check that the fixed type `F` is compatible.
+
+        // We only support fixed types that encode values in [-1,1].
+        // These have a single integer bit.
+        if <T as Fixed>::INT_NBITS != 1 {
+            return Err(FlpError::Encode(format!(
+                "Expected fixed point type with one integer bit, but got {}.",
+                <T as Fixed>::INT_NBITS,
+            )));
+        }
+
+        // Compute number of bits of an entry, and check that an entry fits
+        // into the field.
+        let bits_per_entry: usize = (<T as Fixed>::INT_NBITS + <T as Fixed>::FRAC_NBITS)
+            .try_into()
+            .map_err(|_| FlpError::Encode("Could not convert u32 into usize.".to_string()))?;
+        if !F::valid_integer_bitlength(bits_per_entry) {
+            return Err(FlpError::Encode(format!(
+                "fixed point type bit length ({}) too large for field modulus",
+                bits_per_entry,
+            )));
+        }
+        let fi_bits_per_entry: F::Integer = F::valid_integer_try_from(bits_per_entry)
+            .map_err(|_| FlpError::Encode("Could not create FieldInteger from usize even though size check was successful".to_string()))?;
+
+        // (II) Check that the field is large enough for the norm.
+
+        // Valid norms encoded as field integers lie in [0,2^(2*bits - 2)).
+        let bits_for_norm = 2 * bits_per_entry - 2;
+        if !F::valid_integer_bitlength(bits_for_norm) {
+            return Err(FlpError::Encode(format!(
+                "maximal norm bit length ({}) too large for field modulus",
+                bits_for_norm,
+            )));
+        }
+
+        // In order to compare the actual norm of the vector with the claimed
+        // norm, the field needs to be able to represent all numbers that can
+        // occur during the computation of the norm of any submitted vector,
+        // even if its norm is not bounded by 1. Because of our encoding, an
+        // upper bound to that value is `entries * 2^(2*bits - 2)` (see docs of
+        // compute_norm_of_entries for details). It has to fit into the field.
+        let err = Err(FlpError::Encode(format!(
+            "number of entries ({}) not compatible with field size",
+            entries,
+        )));
+
+        if let Some(val) = (entries as u128).checked_mul(1 << bits_for_norm) {
+            if let Ok(modulus) = u128::try_from(F::modulus()) {
+                if val >= modulus {
+                    return err;
+                }
+            } else {
+                return err;
+            }
+        } else {
+            return err;
+        }
+
+        // Construct the polynomial that computes a part of the norm for a
+        // single vector entry.
+        //
+        // the linear part is 2^n,
+        // the constant part is 2^(2n-2),
+        // the polynomial is:
+        //   p(y) = 2^(2n-2) + -(2^n) * y + 1 * y^2
+        let linear_part = fi_one << (fi_bits_per_entry);
```

Comment:
Curious that clippy doesn't wine here.
```suggestion
        let linear_part = fi_one << fi_bits_per_entry;
```